### PR TITLE
Increase number of connectivity options

### PR DESCRIPTION
**Pull request by squall0gd**
Fixes #

Summary of changes:
- Following configuration flags has been added:
  - `port` - described [here](https://godoc.org/github.com/gocql/gocql#ClusterConfig)
  - `timeout` - described [here](https://godoc.org/github.com/gocql/gocql#ClusterConfig)
  - `connectionTimeout` - described [here](https://godoc.org/github.com/gocql/gocql#ClusterConfig)
  - `initialHostLookup` - described [here](https://godoc.org/github.com/gocql/gocql#ClusterConfig)
  - `ignorePeerAddr`- described [here](https://godoc.org/github.com/gocql/gocql#ClusterConfig)
  - `keyspace` - described [here](https://godoc.org/github.com/gocql/gocql#ClusterConfig)
  - `createKeyspace` - decide that keyspace should be created or not.
- Don't force Keyspace name to `snap`

How to verify it:
- Build and run with external Cassandra instance.

Testing done:
- Tested with production Cassandra

TBD:
- ~~Unit tests~~
- ~~Integration tests~~

A picture of a snapping turtle (not required but encouraged):
snapping turtle

Before change:
If your Cassandra cluster is placed in private network with public endpoint and your administrator give you so little permissions(so you cannot even create your own namespace), you cannot use this plugin.

After change:
This PR brings additional options to configure connection to Cassandra cluster. Following options has been added:
- Flexible keyspace and table names
- Ability to decide that keyspace should be created or not(if your admin doesn't bring you permissions to create your own keyspace and you try to create it, your snap task will fail. Now, you can simply skip keyspace creation)
- Ignore discovering nodes in cluster(if your cluster nodes have got private addresses and public keyspace, than you won't connect to particular nodes in cluster with their private addresses)
- Flexible Timeout and first connection Timeout(in case of big latency on your connection, you will need to increase this two parameters)

### CODE REVIEW

**Comment by squall0gd** [comment]
Resolves #30

**Comment by squall0gd** [comment]
@candysmurf what is the reason of using your cassandra image instead of official one?

File: cassandra/cassandra.go [hunk 1]
```diff
@@ -75,6 +83,41 @@ func (cas *CassandraPublisher) GetConfigPolicy() (*cpolicy.ConfigPolicy, error)
...
+	handleErr(err)
+	serverAddrRule.Description = "Lookup for cluster hosts information, default: true"
+	config.Add(initialHostLookupRule)
+
+	ignorePeerAddrRule, err := cpolicy.NewBoolRule(ignorePeerAddrRuleKey, false, false)
+	handleErr(err)
+	serverAddrRule.Description = "Turn off cluster hosts tracking, default: false"
+	config.Add(ignorePeerAddrRule)
+
+	dontCreateKeyspaceRule, err := cpolicy.NewBoolRule(createKeyspaceRuleKey, false, true)
```
Comment (candysmurf) [inline cassandra/cassandra.go]: createKeyspaceRule

File: cassandra/cassandra.go [hunk 2]
```diff
@@ -223,7 +288,7 @@ func getSslOptions(cfg map[string]ctypes.ConfigValue) *sslOptions {
 
 func handleErr(e error) {
 	if e != nil {
-		log.Fatal(e.Error())
+		log.Fatalf("ABC %s", e.Error())
```
Comment (candysmurf) [inline cassandra/cassandra.go]: what does "ABC" represent here?

File: cassandra/client.go [hunk 3]
```diff
@@ -322,21 +356,17 @@ func addSslOptions(cluster *gocql.ClusterConfig, options *sslOptions) *gocql.Clu
...
-		log.Fatal(err.Error())
-	}
-
-	if err := session.Query(createTableCQL).Exec(); err != nil {
+	if err := session.Query(fmt.Sprintf(createTableCQL, co.keyspace)).Exec(); err != nil {
 		log.Fatal(err.Error())
 	}
 
-	if err := session.Query(createTagTableCQL).Exec(); err != nil {
+	if err := session.Query(fmt.Sprintf(createTagTableCQL, co.keyspace)).Exec(); err != nil {
```
Comment (candysmurf) [inline cassandra/client.go]: should we allow a user to specify the table name similar like the keyspace but default to `snap`?

File: cassandra/client.go [hunk 4]
```diff
@@ -277,21 +287,45 @@ func convert(i interface{}) (interface{}, error) {
...
+	}
+
 	return cluster
 }
 
-func getSession(co clientOptions) *gocql.Session {
-	cluster := createCluster(co.server)
+func createKeyspace(co clientOptions) {
+	cluster := createCluster(co)
+	session, err := cluster.CreateSession()
```
Comment (candysmurf) [inline cassandra/client.go]: using getInstance(), singleton session is desired.

File: cassandra/client.go [hunk 5]
```diff
@@ -277,21 +287,45 @@ func convert(i interface{}) (interface{}, error) {
...
+	if err != nil {
+		log.Fatal(err.Error())
+	}
+	defer session.Close()
```
Comment (candysmurf) [inline cassandra/client.go]: don't close out the session here. Leaving it open

**Comment by candysmurf** [comment]
@squall0gd, using my image was for the medium test to check if the Cassandra is up.

File: cassandra/cassandra.go [hunk 6]
```diff
@@ -38,6 +39,13 @@ const (
 	pluginType = plugin.PublisherPluginType
 
 	serverAddrRuleKey          = "server"
+	portRuleKey                = "port"
+	timeoutRuleKey             = "timeout"
+	connectionTimeoutRuleKey   = "connectionTimeout"
+	initialHostLookupRuleKey   = "initialHostLookup"
+	ignorePeerAddrRuleKey      = "ignoreRuleKey"
+	createKeyspaceRuleKey      = "createKeyspace"
```
Comment (candysmurf) [inline cassandra/cassandra.go]: it seems that we don't need this property. The keyspace is default to `snap`. Keyspace name suffices.

File: cassandra/cassandra.go (hunk 2 above)
Comment (squall0gd) [inline cassandra/cassandra.go]: That was for debug purposes, I've forgot to remove it :/

File: cassandra/client.go (hunk 3 above)
Comment (squall0gd) [inline cassandra/client.go]: we can do this, but I propose to use `metrics` instead of `snap` ;)

File: cassandra/client.go (hunk 4 above)
Comment (squall0gd) [inline cassandra/client.go]: I know, but you need to create keyspace separately. Your configured session isn't able to switch to new keyspace. This case will remove keyspace name from rest of queries. Another approach is to format each query with keyspace name. What do you think?

File: cassandra/cassandra.go (hunk 6 above)
Comment (squall0gd) [inline cassandra/cassandra.go]: My case is that cassandra administrator give me access to cassandra instance with prepared keyspace for me. I cannot create any other keyspace, so even attempt is killing publisher flow.

File: cassandra/client.go (hunk 3 above)
Comment (squall0gd) [inline cassandra/client.go]: What about `tag` table?

File: cassandra/client.go (hunk 4 above)
Comment (candysmurf) [inline cassandra/client.go]: @squall0gd, yes, keep the keyspace in the query.

File: cassandra/client.go (hunk 3 above)
Comment (candysmurf) [inline cassandra/client.go]: @squall0gd, tag table is a by-product. Let us leave it. Also metric as the table name seems fine if we don't want to change too much now.

File: cassandra/cassandra.go [hunk 7]
```diff
@@ -70,51 +79,91 @@ func (cas *CassandraPublisher) GetConfigPolicy() (*cpolicy.ConfigPolicy, error)
...
-	usernameRule, err := cpolicy.NewStringRule(usernameRuleKey, false, "")
+	connectionTimeoutRule, err := cpolicy.NewIntegerRule(connectionTimeoutRuleKey, false, 2)
 	handleErr(err)
-	usernameRule.Description = "Name of a user used to authenticate to Cassandra"
-	config.Add(usernameRule)
+	connectionTimeoutRule.Description = "Initial connection timeout in seconds, default: 2"
+	config.Add(connectionTimeoutRule)
 
-	passwordRule, err := cpolicy.NewStringRule(passwordRuleKey, false, "")
+	dontCreateKeyspaceRule, err := cpolicy.NewBoolRule(createKeyspaceRuleKey, false, true)
```
Comment (squall0gd) [inline cassandra/cassandra.go]: good catch, I've forgot about this one :)

**Comment by PatrykMatyjasek** [comment]
LGTM

File: cassandra/client.go [hunk 8]
```diff
@@ -322,21 +331,23 @@ func addSslOptions(cluster *gocql.ClusterConfig, options *sslOptions) *gocql.Clu
...
-func initializeSession(cluster *gocql.ClusterConfig) *gocql.Session {
+func initializeSession(cluster *gocql.ClusterConfig, co clientOptions) *gocql.Session {
 	session, err := cluster.CreateSession()
 	if err != nil {
 		log.Fatal(err.Error())
 	}
 
-	if err := session.Query(createKeyspaceCQL).Exec(); err != nil {
-		log.Fatal(err.Error())
+	if co.createKeyspace {
```
Comment (candysmurf) [inline cassandra/client.go]: Please remove this createKeyspace flag. It's redundant. As the CQL itself already does this.

File: cassandra/client.go [hunk 9]
```diff
@@ -36,25 +36,40 @@ var (
...
 	ErrInvalidDataType = errors.New("Invalid data type value found - %v")
 
-	createKeyspaceCQL = "CREATE KEYSPACE IF NOT EXISTS snap WITH REPLICATION = {'class': 'SimpleStrategy', 'replication_factor': 1};"
-	createTableCQL    = "CREATE TABLE IF NOT EXISTS snap.metrics (ns  text, ver int, host text, time timestamp, valType text, doubleVal double, strVal text, boolVal boolean, tags map<text,text>, PRIMARY KEY ((ns, ver, host), time),) WITH CLUSTERING ORDER BY (time DESC);"
-	createTagTableCQL = "CREATE TABLE IF NOT EXISTS snap.tags (key  text, val text, time timestamp, ns text, ver int, host text, valType text, doubleVal double, strVal text, boolVal boolean, tags map<text,text>, PRIMARY KEY ((key, val), time),) WITH CLUSTERING ORDER BY (time DESC);"
+	createKeyspaceCQL = "CREATE KEYSPACE IF NOT EXISTS %s WITH REPLICATION = {'class': 'SimpleStrategy', 'replication_factor': 1};"
+	createTableCQL    = "CREATE TABLE IF NOT EXISTS %s.metrics (ns  text, ver int, host text, time timestamp, valType text, doubleVal double, strVal text, boolVal boolean, tags map<text,text>, PRIMARY KEY ((ns, ver, host), time),) WITH CLUSTERING ORDER BY (time DESC);"
```
Comment (candysmurf) [inline cassandra/client.go]: the table name is always `metrics` although the intent is to have user-specified table name but default to `metrics`.

File: cassandra/client.go (hunk 8 above)
Comment (squall0gd) [inline cassandra/client.go]: No it's not :/ if you try to create keyspace and you don't have permissions to your cassandra cluster then your snap task will fail b/c cassandra publisher will panic b/c gocql will exit with panic ;). This flag is simply to handle situation when you know that you don't have permissions.

If you've got permissions to create keyspace, then you're right, but you cannot simply predict if remote cluster give you required permissions or not

File: cassandra/client.go (hunk 9 above)
Comment (squall0gd) [inline cassandra/client.go]: Great catch! Thanks :)

File: cassandra/client.go (hunk 8 above)
Comment (candysmurf) [inline cassandra/client.go]: @squall0gd, user should have privilege set inside Cassandra. we shouldn't use this flag.

Comment (squall0gd) [inline cassandra/client.go]: @candysmurf you cannot expect that your client will have this permissions. And as far as we cannot do that, we need this flag to run tasks for enterprise ready cluster.

Comment (squall0gd) [inline cassandra/client.go]: Here you've got some proof of importance of this flag. 

File `task_w_creating_keyspace.yml` - emulates situation, when this flag isn't needed
File `task_wo_creating_keyspace.yml` - prevents from creating keyspace.

As you can see creating keyspace is killing first task, while second task is up and running.
screencast_05-22-2017_02-12-33 pm

**Comment by candysmurf** [comment]
LGTM, @squall0gd, we can merge this PR after you squash all the commits. thanks.

**Pull request merged**